File: app/src/main/java/com/typingsolutions/kore/setup/ExtendSetupFragment.java
package com.typingsolutions.kore.setup;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.typingsolutions.kore.R;


public class ExtendSetupFragment extends Fragment implements IPasswordProvider {

  private TextInputEditText mEditTextAsEnterPassword;
  private TextInputEditText mEditTextAsRepeatPassword;
  private TextInputEditText mEditTextAsEnterPIM;
  private TextInputEditText mEditTextAsRepeatPIM;
  private TextView mTextViewAsPIMHint;

  private String mBackupText;

  @Nullable
  @Override
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
    View root = inflater.inflate(R.layout.setup_fragment_3, container, false);
    mEditTextAsEnterPassword = (TextInputEditText) root.findViewById(R.id.setuplayout_edittext_passwordenter);
    mEditTextAsRepeatPassword = (TextInputEditText) root.findViewById(R.id.setuplayout_edittext_passwordrepeat);

    mEditTextAsEnterPIM = (TextInputEditText) root.findViewById(R.id.setuplayout_edittext_pimenterextended);
    mEditTextAsRepeatPIM = (TextInputEditText) root.findViewById(R.id.setuplayout_edittext_pimrepeatextended);

    mTextViewAsPIMHint = (TextView) root.findViewById(R.id.setuplayout_textview_currentpim);

    mEditTextAsEnterPassword.addTextChangedListener(new EnableSetupTextWatcher((SetupActivity)getActivity(), this, mEditTextAsRepeatPassword, true));
    mEditTextAsRepeatPassword.addTextChangedListener(new EnableSetupTextWatcher((SetupActivity)getActivity(), this, mEditTextAsEnterPassword));

    mEditTextAsEnterPIM.addTextChangedListener(new EnableSetupTextWatcher((SetupActivity)getActivity(), mEditTextAsRepeatPIM));
    mEditTextAsRepeatPIM.addTextChangedListener(new EnableSetupTextWatcher((SetupActivity)getActivity(), mEditTextAsEnterPIM));

    return root;
  }

  @Override
  public void setPasswords(CharSequence pw1, CharSequence pw2) {
    mEditTextAsEnterPassword.setText(pw1);
    mEditTextAsRepeatPassword.setText(pw2);
  }

  @Override
  public void cleanUp() {
    SetupActivity activity = (SetupActivity) getActivity();

    activity.clearText(mEditTextAsEnterPassword);
    activity.clearText(mEditTextAsRepeatPassword);

    mEditTextAsEnterPassword.clearComposingText();
    mEditTextAsRepeatPassword.clearComposingText();

    mEditTextAsEnterPassword = null;
    mEditTextAsRepeatPassword = null;

    activity.clearText(mEditTextAsEnterPIM);
    activity.clearText(mEditTextAsRepeatPIM);

    mEditTextAsEnterPIM.clearComposingText();
    mEditTextAsRepeatPIM.clearComposingText();

    mEditTextAsEnterPIM = null;
    mEditTextAsRepeatPIM = null;
  }

  @Override
  public CharSequence getPassword1() {
    return mEditTextAsEnterPassword.getText();
  }

  @Override
  public CharSequence getPassword2() {
    return mEditTextAsRepeatPassword.getText();
  }

  CharSequence getPIM1() {
    return mEditTextAsEnterPIM.getText();
  }

  CharSequence getPIM2() {
    return mEditTextAsRepeatPIM.getText();
  }

  void setCurrentPIM() {
    if (mBackupText == null) {
      mBackupText = mTextViewAsPIMHint.getText().toString();
    }

    SetupActivity activity = (SetupActivity) getActivity();

    int calcPim = activity.mKoreApplication.calculatePIM(mEditTextAsEnterPassword.getText().toString());
    mTextViewAsPIMHint.setText(mBackupText.replace("${_pim_}", Integer.toString(calcPim)));
  }
}

File: app/src/main/java/com/typingsolutions/kore/setup/ExtendSetupFragment.kt
package com.typingsolutions.kore.setup

import android.os.Bundle
import android.support.design.widget.TextInputEditText
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.typingsolutions.kore.R


class ExtendSetupFragment : Fragment(), IPasswordProvider {

    private lateinit var mEditTextAsEnterPassword: TextInputEditText
    private lateinit var mEditTextAsRepeatPassword: TextInputEditText
    private lateinit var mEditTextAsEnterPIM: TextInputEditText
    private lateinit var mEditTextAsRepeatPIM: TextInputEditText
    private var mTextViewAsPIMHint: TextView? = null

    private var mBackupText: String? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater!!.inflate(R.layout.setup_fragment_3, container, false)
        mEditTextAsEnterPassword = root.findViewById(R.id.setuplayout_edittext_passwordenter) as TextInputEditText
        mEditTextAsRepeatPassword = root.findViewById(R.id.setuplayout_edittext_passwordrepeat) as TextInputEditText

        mEditTextAsEnterPIM = root.findViewById(R.id.setuplayout_edittext_pimenterextended) as TextInputEditText
        mEditTextAsRepeatPIM = root.findViewById(R.id.setuplayout_edittext_pimrepeatextended) as TextInputEditText

        mTextViewAsPIMHint = root.findViewById(R.id.setuplayout_textview_currentpim) as TextView

        mEditTextAsEnterPassword.addTextChangedListener(EnableSetupTextWatcher(activity as SetupActivity, this, mEditTextAsRepeatPassword, true))
        mEditTextAsRepeatPassword.addTextChangedListener(EnableSetupTextWatcher(activity as SetupActivity, this, mEditTextAsEnterPassword))

        mEditTextAsEnterPIM.addTextChangedListener(EnableSetupTextWatcher(activity as SetupActivity, mEditTextAsRepeatPIM))
        mEditTextAsRepeatPIM.addTextChangedListener(EnableSetupTextWatcher(activity as SetupActivity, mEditTextAsEnterPIM))

        return root
    }

    override fun setPasswords(pw1: CharSequence, pw2: CharSequence) {
        mEditTextAsEnterPassword.setText(pw1)
        mEditTextAsRepeatPassword.setText(pw2)
    }

    override fun cleanUp() {
        val activity = activity as SetupActivity

        activity.clearText(mEditTextAsEnterPassword)
        activity.clearText(mEditTextAsRepeatPassword)

        mEditTextAsEnterPassword.clearComposingText()
        mEditTextAsRepeatPassword.clearComposingText()

        activity.clearText(mEditTextAsEnterPIM)
        activity.clearText(mEditTextAsRepeatPIM)

        mEditTextAsEnterPIM.clearComposingText()
        mEditTextAsRepeatPIM.clearComposingText()
    }

    override fun getPassword1(): CharSequence? {
        return mEditTextAsEnterPassword.text
    }

    override fun getPassword2(): CharSequence? {
        return mEditTextAsRepeatPassword.text
    }

    internal val pim1: CharSequence
        get() = mEditTextAsEnterPIM.text

    internal val pim2: CharSequence
        get() = mEditTextAsRepeatPIM.text

    internal fun setCurrentPIM() {
        if (mBackupText == null) {
            mBackupText = mTextViewAsPIMHint?.text.toString()
        }

        val activity = activity as SetupActivity

        val calcPim = activity.mKoreApplication.calculatePIM(mEditTextAsEnterPassword.text.toString())
        mTextViewAsPIMHint?.text = mBackupText?.replace("\${_pim_}", Integer.toString(calcPim))
    }
}
